src/data.table.h

```diff
@@ -47,6 +47,9 @@ typedef R_xlen_t RLEN;
 #define NA_INTEGER64  INT64_MIN
 #define MAX_INTEGER64 INT64_MAX
 
+// for use with CPLXSXP, no macro provided by R internals
+#define ISNAN_COMPLEX(x) (ISNAN(x.r) || ISNAN(x.i)) // TRUE if either real or imaginary component is NA or NaN
```

Comment:
A few other places this could be used if we're going to add it:

```
src/coalesce.c:128:        if (ISNAN(tt.r) && ISNAN(tt.i)) continue;
src/coalesce.c:134:    const bool final = !ISNAN(finalVal.r) && !ISNAN(finalVal.i);
src/coalesce.c:138:      if (!ISNAN(val.r) && !ISNAN(val.i)) continue;
src/coalesce.c:139:      int j=0; while (ISNAN(val.r) && ISNAN(val.i) && j<k) val=((Rcomplex *)valP[j++])[i];
src/coalesce.c:140:      if (!ISNAN(val.r) || !ISNAN(val.i)) xP[i]=val; else if (final) xP[i]=finalVal;
src/frank.c:59:      for (int j=0; j<n; ++j) ians[j] |= (ISNAN(COMPLEX(v)[j].r) || ISNAN(COMPLEX(v)[j].i));
src/frank.c:195:      while (j < n && !ISNAN(COMPLEX(v)[j].r) && !ISNAN(COMPLEX(v)[j].i)) j++;
src/gsumm.c:320:          if (ISNAN(elem.r) && ISNAN(elem.i)) my_anyNA = true;
src/gsumm.c:327:          if (ISNAN(elem.r) && ISNAN(elem.i)) my_anyNA = true;
src/gsumm.c:557:            if (!ISNAN(elem.r)) _ans[my_low[i]].r += elem.r;
src/gsumm.c:642:      if (ISNAN(xd[ix].r) || ISNAN(xd[ix].i)) continue;  // || otherwise we'll need two counts in two c's too?
src/assign.c:1039:      else          BODY(double,  REAL, double, ISNAN(val)?(im=NA_REAL,NA_REAL):(im=0.0,val),         td[i].r=cval;td[i].i=im)
```

Reply:
It strikes me as a bit odd that there aren't macros for these in Rinternals.h - I'll how R-devel feels about adding them to base R